linera-core/src/client/mod.rs

```diff
@@ -1132,79 +1180,111 @@ where
         // Retrieve newly received certificates from this validator.
         let query = ChainInfoQuery::new(chain_id).with_received_log_excluding_first_n(tracker);
         let info = remote_node.handle_chain_info_query(query).await?;
-        let mut certificates: Vec<Certificate> = Vec::new();
-        let mut new_tracker = tracker;
+        let mut tracker = MultichainTracker::new(tracker, info.requested_received_log.clone());
+        let remote_node_chains_view = info.requested_received_log.into_iter().fold(
+            BTreeMap::<ChainId, Vec<BlockHeight>>::new(),
+            |mut chain_to_info, entry| {
+                chain_to_info
+                    .entry(entry.chain_id)
+                    .or_default()
+                    .push(entry.height);
+                chain_to_info
+            },
+        );
 
-        for entry in info.requested_received_log {
-            let query = ChainInfoQuery::new(entry.chain_id)
-                .with_sent_certificate_hashes_in_range(BlockHeightRange::single(entry.height));
+        // Collect all certificates that have been successfully processed.
+        let mut certificates: Vec<Certificate> = Vec::new();
 
-            let local_response = self
-                .client
-                .local_node
-                .handle_chain_info_query(query.clone())
-                .await
-                .map_err(|error| NodeError::LocalNodeQuery {
-                    error: error.to_string(),
-                })?
-                .info;
-            if !local_response.requested_sent_certificate_hashes.is_empty() {
-                // We've already processed incoming messages for this certificate.
-                new_tracker += 1;
-                continue;
+        for (chain_id, mut block_batch) in remote_node_chains_view {
+            block_batch.sort();
+
+            'local: loop {
+                let block_height = block_batch.first().copied().unwrap();
+                let local_query = ChainInfoQuery::new(chain_id);
+                let local_response = self.local_query(local_query).await?;
+                if !local_response
+                    .info
+                    .requested_sent_certificate_hashes
```

Comment:
But we didn't request any sent certificate hashes.

Reply:
🙈